WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/PostListAdapter.kt

```diff
@@ -134,43 +116,45 @@ class PostListAdapter(
         onPostButtonClickListener = listener
     }
 
-    private fun getItem(position: Int): PostModel? {
-        return if (isValidPostPosition(position)) {
-            posts[position]
-        } else null
-    }
-
-    private fun isValidPostPosition(position: Int): Boolean {
-        return position >= 0 && position < posts.size
-    }
-
     override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
         super.onAttachedToRecyclerView(recyclerView)
         this.recyclerView = recyclerView
     }
 
     override fun getItemViewType(position: Int): Int {
-        return if (position == posts.size) {
-            VIEW_TYPE_ENDLIST_INDICATOR
-        } else VIEW_TYPE_POST
+        return when {
+            position == (itemCount - 1) -> VIEW_TYPE_ENDLIST_INDICATOR
+            listManager?.getItem(position) != null -> VIEW_TYPE_POST
+            else -> VIEW_TYPE_LOADING
+        }
     }
 
     override fun getItemCount(): Int {
-        return if (posts.size == 0) {
-            0
-        } else {
-            posts.size + 1 // +1 for the endlist indicator
+        listManager?.let {
+            return if (it.size == 0) {
+                0
+            } else {
+                it.size + 1 // +1 for the endlist indicator
+            }
         }
+        return 0
     }
 
     override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
-        return if (viewType == VIEW_TYPE_ENDLIST_INDICATOR) {
-            val view = layoutInflater.inflate(R.layout.endlist_indicator, parent, false)
-            view.layoutParams.height = endlistIndicatorHeight
-            EndListViewHolder(view)
-        } else {
-            val view = layoutInflater.inflate(R.layout.post_cardview, parent, false)
-            PostViewHolder(view)
+        return when (viewType) {
+            VIEW_TYPE_ENDLIST_INDICATOR -> {
+                val view = layoutInflater.inflate(R.layout.endlist_indicator, parent, false)
+                view.layoutParams.height = endlistIndicatorHeight
+                EndListViewHolder(view)
+            }
+            VIEW_TYPE_LOADING -> {
+                val view = layoutInflater.inflate(R.layout.post_cardview_skeleton, parent, false)
+                LoadingViewHolder(view)
+            }
+            else -> {
```

Comment:
In order to follow Fail-Fast principle we may consider explicitly specifying VIEW_TYPE_POST and throw an exception in the else branch.

Reply:
Agreed, fixed in 8719cdf4fd7a263d114ee14e44f6402ffb8fec85.

The changeset looks bigger than it actually is in that commit due to indentation change. 😢 

That commit also updates the FluxC hash. It includes some important changes made in FluxC's `develop`  branch. If you already installed the app using this branch, you'll need to do a clean installation as the migration will fail. (had to change it due to a merge conflict) /cc @kwonye